File: fioincline/lvovich/incline.py
"""Склонение ФИО — порт index.ts / incline.go (nodkz/lvovich)."""

from . import types as T
from .rules import incline_by_rules
from .gender import Person, get_gender_const
from .rules_data import firstname_rules, lastname_rules, middlename_rules


def resolve_declension(declension):
    if declension is None or declension == "":
        return T.ACCUSATIVE
    return T.get_declension_const(declension)


def incline_firstname(str_, declension, gender):
    if gender == T.GENDER_NULL:
        from .gender import get_fg
        gender = get_fg(str_)
    return incline_by_rules(str_, resolve_declension(declension), gender, firstname_rules)


def incline_lastname(str_, declension, gender):
    if gender == T.GENDER_NULL:
        from .gender import get_lg
        gender = get_lg(str_)
    return incline_by_rules(str_, resolve_declension(declension), gender, lastname_rules)


def incline_middlename(str_, declension, gender):
    if gender == T.GENDER_NULL:
        from .gender import get_mg
        gender = get_mg(str_)
    return incline_by_rules(str_, resolve_declension(declension), gender, middlename_rules)


class InclineOut:
    __slots__ = ("first_name", "sur_name", "second_name", "gender", "initials")

    def __init__(self):
        self.first_name = ""
        self.sur_name = ""
        self.second_name = ""
        self.gender = ""
        self.initials = ""


def initials(p):
    res = ""
    if p.first_name != "":
        first = p.first_name.strip()
        if first != "":
            res += first[0].upper() + "."
    if p.second_name != "":
        second = p.second_name.strip()
        if second != "":
            res += second[0].upper() + "."
    return res


def incline(p, declension, format_=""):
    from .gender import get_gender
    gender = get_gender(p)
    res = InclineOut()
    res.gender = gender
    gc = get_gender_const(p)

    if p.first_name != "":
        res.first_name = incline_firstname(p.first_name.strip(), declension, gc)
    if p.sur_name != "":
        res.sur_name = incline_lastname(p.sur_name.strip(), declension, gc)
    if p.second_name != "":
        res.second_name = incline_middlename(p.second_name.strip(), declension, gc)

    if format_ == "initials":
        res.initials = initials(p)
    return res

File: internal/lvovich/incline.go
// Склонение ФИО — порт index.ts
// библиотеки nodkz/lvovich (https://github.com/nodkz/lvovich), MIT.
package lvovich

import (
	"strings"
	"unicode"
	"unicode/utf8"
)

// resolveDeclension — приведение переданного падежа к константе.
// undefined/пустое значение трактуется как винительный (по умолчанию оригинала).
func resolveDeclension(declension interface{}) DeclenType {
	if declension == nil {
		return Accusative
	}
	switch v := declension.(type) {
	case string:
		if v == "" {
			return Accusative
		}
	case DeclenType:
		if v == DeclenNull {
			return Accusative
		}
	}
	return GetDeclensionConst(declension)
}

// InclineFirstname — склонение имени.
func InclineFirstname(str string, declension interface{}, gender GenderConst) string {
	if gender == GenderNull {
		gender = GetFG(str)
	}
	return inclineByRules(str, resolveDeclension(declension), gender, firstnameRules)
}

// InclineLastname — склонение фамилии.
func InclineLastname(str string, declension interface{}, gender GenderConst) string {
	if gender == GenderNull {
		gender = GetLG(str)
	}
	return inclineByRules(str, resolveDeclension(declension), gender, lastnameRules)
}

// InclineMiddlename — склонение отчества.
func InclineMiddlename(str string, declension interface{}, gender GenderConst) string {
	if gender == GenderNull {
		gender = GetMG(str)
	}
	return inclineByRules(str, resolveDeclension(declension), gender, middlenameRules)
}

// InclineOut — результат склонения ФИО.
type InclineOut struct {
	FirstName  string
	SurName    string
	SecondName string
	Gender     string
	// Initials заполняется только в формате initials.
	Initials string
}

// initials — инициалы вида "И.И." (первая буква имени/отчества с точкой).
func initials(p Person) string {
	res := ""
	if p.FirstName != "" {
		first := strings.TrimSpace(p.FirstName)
		if first != "" {
			r, _ := utf8.DecodeRuneInString(first)
			res += string(unicode.ToUpper(r)) + "."
		}
	}
	if p.SecondName != "" {
		second := strings.TrimSpace(p.SecondName)
		if second != "" {
			r, _ := utf8.DecodeRuneInString(second)
			res += string(unicode.ToUpper(r)) + "."
		}
	}
	return res
}

// Incline — склонение ФИО по падежам.
// format: "full" (по умолчанию) или "initials".
func Incline(p Person, declension interface{}, format string) InclineOut {
	gender := GetGender(p)
	res := InclineOut{Gender: gender}
	gc := GetGenderConst(gender)

	if p.FirstName != "" {
		res.FirstName = InclineFirstname(strings.TrimSpace(p.FirstName), declension, gc)
	}
	if p.SurName != "" {
		res.SurName = InclineLastname(strings.TrimSpace(p.SurName), declension, gc)
	}
	if p.SecondName != "" {
		res.SecondName = InclineMiddlename(strings.TrimSpace(p.SecondName), declension, gc)
	}

	if format == "initials" {
		res.Initials = initials(p)
	}
	return res
}